src/util/utils.js

```diff
@@ -15,37 +16,61 @@ const BLOCK_HOST_NAMES_LIST = BLOCK_HOST_NAMES.split(',');
 const LOCAL_HOST_NAMES_LIST = ['localhost', '127.0.0.1', '[::]', '[::1]'];
 const LOCALHOST_OCTET = '127.';
 const RECORD_TYPE_A = 4; // ipv4
+const DNS_CACHE_ENABLED = process.env.DNS_CACHE_ENABLED === 'true';
+const DNS_CACHE_TTL = process.env.DNS_CACHE_TTL ? parseInt(process.env.DNS_CACHE_TTL, 10) : 300;
+const dnsCache = new NodeCache({
+  useClones: false,
+  stdTTL: DNS_CACHE_TTL,
+  checkperiod: DNS_CACHE_TTL,
+});
+
+const fetchResolvedIp = async (hostname) => {
+  // ex: [{ address: '108.157.0.0', ttl: 600 }]
+  const addresses = await resolver.resolve4(hostname, { ttl: true });
+  return addresses.length > 0 ? addresses[0] : {};
+};
 
 const staticLookup = (transformationTags) => async (hostname, _, cb) => {
-  let ips;
+  let ip;
   const resolveStartTime = new Date();
+  let dnsHit = false;
   try {
-    ips = await resolver.resolve4(hostname);
+    if (DNS_CACHE_ENABLED) {
+      ip = dnsCache.get(hostname);
+      if (ip !== undefined) {
+        dnsHit = true;
+      } else {
+        const { address, ttl } = await fetchResolvedIp(hostname);
+        ip = address;
+        dnsCache.set(hostname, ip, ttl || DNS_CACHE_TTL);
+      }
+    } else {
+      const { address } = await fetchResolvedIp(hostname);
+      ip = address;
+    }
```

Comment:
What do you think if we re-write all of this with a decorator pattern ?

```javascript
@cache
function hostnameToIP(hostname) {
}
```

Inside the cache decorator we can read the env values and then fetch the IP from cache or resolve it actually.